yazi-plugin/src/external/highlighter.rs

```diff
@@ -56,51 +59,235 @@ impl Highlighter {
+		let mut lines_handled = 0;
+		for mut long_line in long_lines {
+			if long_line.is_empty() || lines_handled > skip + area.height as usize {
+				break;
+			}
+			Self::replace_tabs_with_spaces(&mut long_line, PREVIEW.tab_size as usize);
+			for line in Self::chunk_by_width(long_line, area.width as usize) {
+				let mut line = line.to_vec();
+				lines_handled += 1;
+				let must_break = Self::handle_single_line(
+					lines_handled,
+					skip,
+					*plain,
+					area.height as usize,
+					&mut line,
+					&mut before,
+					&mut after,
+				);
+				if must_break {
+					break;
+				}
 			}
+		}
+
+		let no_more_scroll = lines_handled < skip + area.height as usize;
+		if skip > 0 && no_more_scroll {
+			return Err(PeekError::Exceed(lines_handled.saturating_sub(area.height as usize)));
+		}
 
-			if i > skip {
-				after.push(String::from_utf8_lossy(&buf).into_owned());
-			} else if !plain {
-				before.push(String::from_utf8_lossy(&buf).into_owned());
+		Ok((before, after))
+	}
+
+	async fn before_after(
+		skip: usize,
+		area: Rect,
+		plain: &mut bool,
+		mut reader: BufReader<File>,
+	) -> Result<(Vec<String>, Vec<String>), PeekError> {
+		let mut before = Vec::with_capacity(if *plain { 0 } else { skip });
+		let mut after = Vec::with_capacity(area.height as usize);
+
+		let mut lines_handled = 0;
+		let mut buf = vec![];
+		while reader.read_until(b'\n', &mut buf).await.is_ok() {
+			lines_handled += 1;
+			if !*plain && buf.len() > MAX_LINE_BYTES_TO_PLAINTEXT_FALLBACK {
+				*plain = true;
+				drop(mem::take(&mut before));
+			}
+			let must_break = Self::handle_single_line(
+				lines_handled,
+				skip,
+				*plain,
+				area.height as usize,
+				&mut buf,
+				&mut before,
+				&mut after,
+			);
+			if must_break {
+				break;
 			}
 			buf.clear();
 		}
 
-		if skip > 0 && i < skip + limit {
-			return Err(PeekError::Exceed(i.saturating_sub(limit)));
+		let no_more_scroll = lines_handled < skip + area.height as usize;
+		if skip > 0 && no_more_scroll {
+			return Err(PeekError::Exceed(lines_handled.saturating_sub(area.height as usize)));
 		}
 
-		if plain {
-			let indent = " ".repeat(PREVIEW.tab_size as usize);
-			Ok(Text::from(after.join("").replace('\t', &indent)))
-		} else {
-			Self::highlight_with(before, after, syntax.unwrap()).await
+		Ok((before, after))
+	}
+
+	fn handle_single_line(
+		lines_handled: usize,
+		skip: usize,
+		plain: bool,
+		limit: usize,
+		line: &mut Vec<u8>,
+		before: &mut Vec<String>,
+		after: &mut Vec<String>,
+	) -> bool {
+		if line.is_empty() || lines_handled > skip + limit {
+			return true;
+		}
+
+		if line.ends_with(b"\r\n") {
+			line.pop();
+			line.pop();
+			line.push(b'\n');
+		} else if !line.ends_with(b"\n") {
+			line.push(b'\n')
+		}
+
+		let text = String::from_utf8_lossy(line).into_owned();
+		if lines_handled > skip {
+			after.push(text);
+		} else if !plain {
+			before.push(text);
+		}
+		false
+	}
+
+	fn chunk_by_width(line: Vec<u8>, width: usize) -> Vec<Vec<u8>> {
```

Comment:
```suggestion
	fn chunk_by_width(line: Vec<u8>, width: usize) -> Vec<String> {
```

Perhaps `String` is a better return type as we need to convert `Vec<u8>` to `String` anyway in order to break the content.

Reply:
Changed return type to Vec<String>.

If you see there is a big optimization opportunity which I don't see, please feel free to commit, I don't mind